pkg/ext/stores/useractivity/store.go

```diff
@@ -0,0 +1,199 @@
+package useractivity
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	ext "github.com/rancher/rancher/pkg/apis/ext.cattle.io/v1"
+	v3Legacy "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	v3 "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/rancher/rancher/pkg/wrangler"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/registry/rest"
+)
+
+const (
+	UserActivityNamespace = "cattle-useractivity-data"
+	tokenUserId           = "authn.management.cattle.io/token-userId"
+	SingularName          = "useractivity"
+	PluralName            = "useractivities"
+)
+
+// +k8s:openapi-gen=false
+// +k8s:deepcopy-gen=false
+type Store struct {
+	tokenController v3.TokenController
+}
+
+var GV = schema.GroupVersion{
+	Group:   "ext.cattle.io",
+	Version: "v1",
+}
+
+var GVK = schema.GroupVersionKind{
+	Group:   GV.Group,
+	Version: GV.Version,
+	Kind:    "UserActivity",
+}
+var GVR = schema.GroupVersionResource{
+	Group:    GV.Group,
+	Version:  GV.Version,
+	Resource: PluralName,
+}
+
+func NewFromWrangler(wranglerCtx *wrangler.Context) *Store {
+	return &Store{
+		tokenController: wranglerCtx.Mgmt.Token(),
+	}
+}
+
+// GroupVersionKind implements [rest.GroupVersionKindProvider]
+func (t *Store) GroupVersionKind(_ schema.GroupVersion) schema.GroupVersionKind {
+	return GVK
+}
+
+// NamespaceScoped implements [rest.Scoper]
+func (t *Store) NamespaceScoped() bool {
+	return false
+}
+
+// GetSingularName implements [rest.SingularNameProvider]
+func (t *Store) GetSingularName() string {
+	return SingularName
+}
+
+// New implements [rest.Storage]
+func (t *Store) New() runtime.Object {
+	obj := &ext.UserActivity{}
+	obj.GetObjectKind().SetGroupVersionKind(GVK)
+	return obj
+}
+
+// Destroy implements [rest.Storage]
+func (t *Store) Destroy() {
+}
+
+// Create implements [rest.Creator]
+func (uas *Store) Create(ctx context.Context,
+	obj runtime.Object,
+	createValidation rest.ValidateObjectFunc,
+	options *metav1.CreateOptions) (runtime.Object, error) {
+	if createValidation != nil {
+		err := createValidation(ctx, obj)
+		if err != nil {
+			return obj, err
+		}
+	}
+
+	// retrieving useractivity object from raw data
+	objUserActivity, ok := obj.(*ext.UserActivity)
+	if !ok {
+		var zeroUA *ext.UserActivity
+		return nil, apierrors.NewInternalError(fmt.Errorf("expected %T but got %T", zeroUA, objUserActivity))
+	}
+	// retrieve token information
+	token, err := uas.tokenController.Get(objUserActivity.Spec.TokenID, metav1.GetOptions{})
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to get token %s: %v", objUserActivity.Spec.TokenID, err))
```

Comment:
Note: If the `err` here is actually a `not found error`, then this is a `BadRequest`, not an internal error.

Reply:
I believe that you misunderstood me here.
I did not ask to unconditionally change the NewInternal to NewBadRequest.
I asked to check if the err you is a a `not found`, or not, and then use NewInternal or NewBad accordingly.
I.e. I expected the code to become, roughly
```
if err != nil {
    if is-not-found (err) {
        return nil, NewBadRequest ...
    } else {
        return nil, NewInternalError ...
   }
}
```
Apologies for my ambiguous comment